FeedbackSessionsLogicTest: use a cleaner method to compare the csv results #4629

jkt001 opened the pull request:
Fixes #4629 (and as a consequent of this method update, fixes #4630)

However, there is this one "naughty" test case. 

Eclipse does not show any differences between the compared results, but the test fails. If you look carefully, there seems to be no differences (refer to line 1187 of FeedbackSessionLogicTest.java). This is currently commented and if so, all testcases are passed. 

Here is the screenshot of said test. It doesn't show everything, but there are no indication of mistakes whatsoever (at least what I saw in my Eclipse).

<img width="1763" alt="screen shot 2016-01-16 at 15 46 32" src="https://cloud.githubusercontent.com/assets/8586215/12371343/3c063f06-bc6b-11e5-9485-ae0bdd25bd99.png">

<img width="1759" alt="screen shot 2016-01-16 at 16 09 05" src="https://cloud.githubusercontent.com/assets/8586215/12371351/92c18ae4-bc6b-11e5-935c-070b6df04cf0.png">

jkt001 commented on src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java:
```diff
@@ -1127,39 +1128,64 @@ public void testGetFeedbackSessionResultsSummaryAsCsv() throws Exception {
+            ""
+        };
+        
+      // assertEquals(StringUtils.join(expected, Const.EOL), export);        
```
This test is problematic. However, as the screenshot above suggested, there are no "visual differences" between the actual and expected. Anyone can help on this? thanks

damithc replied:
A missing CR or LF character? (the use of CR and LF to indicate line ending is different between Mac and Windows).

jkt001 replied:
hmm I added a Carriage Return <code>\r</code> and it passed the testcase. Let me update the PR and see if it is the desired changes. However, I don't know the implications this will have on Windows devices.

jkt001 commented:
Thanks @damithc for the heads-up.

I have updated the file (uncommented the test case, added a <CR> <code>\r</code> to line 1187. This passed the test on a Mac. However, I don't know its effect on Windows devices.

If the previous test (before adding <code>\r</code>) passed on Windows, then I guess issue #4630 still persist after all, even though the method of testing is updated.

LowWeiLin commented:
The test fails on windows, due to the ```\r```
We will have to hunt down the location ```\n or \r\n``` is used instead of ```Const.EOL```

jkt001 commented:
Ok.

Should I revert back the changes (i.e. remove the <code>\r</code>) for this issue, since we have #4630 for that problem (hence another PR to resolve it), or do I need to fix it through this PR as well?

I'm fine with either option since I am currently looking to resolve it without having to add <code>\r</code> to that specific line.

LowWeiLin commented:
Try to figure out how to fix it? It can be done in this PR. #4630 is slightly different as it is specific to rank question type.
Figure out where the test case comes from first to understand why the test fails.

jkt001 commented:
It is fixed with my latest push. 

Issue #4630 is fixed since the initial modification of the file (i.e. from my first commit). But then this issue is a new one (the same kind of error as #4630 but in a different location).

Currently, as of my latest push, all test-cases are passed in a Mac.

I do hope now it passes the tests on Windows.

LowWeiLin commented:
The test passes now, but the fix does not look right..
The data for this test case is found in typicalDatabundle.json, and the response is:
```
"response3ForQ2S1C1":{
            "feedbackSessionName":"First feedback session",
            "courseId":"idOfTypicalCourse1",
            "feedbackQuestionId":"2",
            "feedbackQuestionType":"TEXT",
            "giverEmail":"student3InCourse1@gmail.tmt",
            "recipientEmail":"student2InCourse1@gmail.tmt",
            "responseMetaData":{
                "value":"Response from student 3 \"to\" student 2.\r\nMultiline test."
            },
            "giverSection":"Section 1",
            "recipientSection":"Section 1" 
            
        },
```
You can see that ```\r\n``` is hard-coded in the response. 
I would think that a more understandable fix would be to hard-code the expected result as well.
Replacing Const.EOL does not really make sense to me.

```
"\"Team 1.1\",\"student3 In Course1\",\"Course1\",\"student3InCourse1@gmail.tmt\",\"Team 1.1\",\"student2 In Course1\",\"Course1\",\"student2InCourse1@gmail.tmt\",\"Response from student 3 \"\"to\"\" student 2.\r\nMultiline test.\"",
```

jkt001 commented:
OK this works. No wonder why the error shows up now instead of before.

I have changed how the expected file is shown in the preceding comments as well to reflect how the response is actually is. This is similar to the hard-coded <code>\n</code> in line 2037-2040, which was the cause of issue #4630.

All tests now pass on my Mac. Let me know if this requires anymore changes.

LowWeiLin commented:
Great! InstructorFeedbackResultsDownloadActionTest has csv comparisons that needs to be changed as well.
Once that is done and the tests pass, this PR should be done.

jkt001 commented:
@LowWeiLin 

From what I have observed in InstructorFeedbackResultsDownloadActionTest, the csv comparison result is exactly the same as in FeedbackSessionsLogicTest. 

There are three tests that have this same pattern (compare csv results) in the file InstructorFeedbackResultsDownloadActionTest, from line 120-235.

That one currently only tests the first few lines of the CSV result, with the content itself is explicitly mentioned being tested in FeedbackSessionsLogicTest.

Hence, If I modify the three tests in that file to be similar to what is being done in FeedbackSessionsLogicTest, we would have three identical tests in InstructorFeedbackResultsDownloadActionTest as in FeedbackSessionsLogicTest (since we have no choice but to include test for the file contents as well, which is already being tested in FeedbackSessionsLogicTest).

I can do it if there is no problem with duplicate tests on two different files, if needed. 

Even better, since they are tested in FeedbackSessionsLogicTest, we might not need that test in InstructorFeedbackResultsDownloadActionTest after all and remove them altogether. This can be done in a separate issue and PR, if needed.

This is just my opinion by the way. Just inform me on which is the better option, and I would be glad to implement it.

damithc commented:
Does this involve comparing content of a csv file in the repo? In which case, its line endings depend on which OS was used to create the file before it was committed to the repo, not the OS used for testing now.

LowWeiLin commented:
The csv file is not in the repo. It is part of a FileDownloadResult.

There is a difference between FeedbackSessionsLogicTest and InstructorFeedbackResultsDownloadActionTest.

FeedbackSessionsLogicTest tests the content generated by the logic classes, while InstructorFeedbackResultsDownloadActionTest tests if the csv file can be downloaded (through FileDownloadResult) and if the content exists and is correct (i suppose that's why there is only a partial test of the content)
So, we should not remove InstructorFeedbackResultsDownloadActionTest.
The testing can still be made cleaner like in FeedbackSessionsLogicTest, by asserting that the file content begins/starts with some expected result.

jkt001 commented:
I have updated the tests in InstructorFeedbackResultsDownloadActionTest as well. 

Do let me know if there's anything I should change / add again.

LowWeiLin commented:
Looks good, will merge this soon.
